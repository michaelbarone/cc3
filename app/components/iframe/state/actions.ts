import { IframeAction, IframeStatus } from "@/app/types/iframe";

export function setIframeStatus(urlId: string, status: IframeStatus): IframeAction {
  return {
    type: "SET_STATUS",
    payload: { urlId, status },
  };
}

export function setIframeError(urlId: string, error: string | null): IframeAction {
  return {
    type: "SET_ERROR",
    payload: { urlId, error },
  };
}

export function updateIframeActivity(urlId: string): IframeAction {
  return {
    type: "UPDATE_ACTIVITY",
    payload: { urlId },
  };
}

export function resetIframe(urlId: string): IframeAction {
  return {
    type: "RESET_IFRAME",
    payload: { urlId },
  };
}

export function setIdleTimeout(urlId: string, timeout: number): IframeAction {
  return {
    type: "SET_IDLE_TIMEOUT",
    payload: { urlId, timeout },
  };
}

export function setUrl(urlId: string, url: string, urlMobile: string | null): IframeAction {
  return {
    type: "SET_URL",
    payload: { urlId, url, urlMobile },
  };
}
