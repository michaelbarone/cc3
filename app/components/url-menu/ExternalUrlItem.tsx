import { Url } from "@/app/lib/types";
import { getEffectiveUrl } from "@/app/lib/utils/iframe-utils";
import OpenInNewIcon from "@mui/icons-material/OpenInNew";
import {
  Box,
  Button,
  ListItemButton,
  ListItemText,
  Theme,
  Tooltip,
  useMediaQuery,
} from "@mui/material";
import { memo, useMemo } from "react";

interface ExternalUrlItemProps {
  url: Url;
  tooltipText: string;
  menuPosition: "top" | "side";
  theme: Theme;
}

const ExternalUrlItem = memo(function ExternalUrlItem({
  url,
  tooltipText,
  menuPosition,
  theme,
}: ExternalUrlItemProps) {
  // Check if we're on mobile
  const isMobile = useMediaQuery(theme.breakpoints.down("md"));

  // Simple click handler that opens the URL in a new tab
  const handleClick = () => {
    // Get the effective URL based on localhost settings and device
    const effectiveUrl = url.isLocalhost
      ? getEffectiveUrl(
          {
            id: url.id,
            url: url.url,
            urlMobile: url.urlMobile,
            isLocalhost: true,
            port: url.port || null,
            path: url.path || null,
            localhostMobilePath: url.localhostMobilePath || null,
            localhostMobilePort: url.localhostMobilePort || null,
          },
          isMobile,
        )
      : isMobile && url.urlMobile
        ? url.urlMobile
        : url.url;

    window.open(effectiveUrl, "_blank", "noopener,noreferrer");
  };

  // For top menu, use Button-based styling
  if (menuPosition === "top") {
    const styles = useMemo(
      () => ({
        iconStyles: {
          width: 24,
          height: 24,
          objectFit: "contain" as const,
          marginRight: url.title && url.iconPath ? 1 : 0,
        },
        boxStyles: {
          position: "relative" as const,
          display: "flex",
          alignItems: "center",
          justifyContent: "center",
          width: "100%",
        },
        buttonStyles: {
          height: 50, // Fixed height for consistency
          minHeight: 50, // Ensure minimum height is the same
          minWidth: 50,
          lineHeight: "36px", // Match line height to button height
          px: 0, // Slightly more horizontal padding
          mx: 0.5,
          textTransform: "none",
          borderRadius: 1,
          color: theme.palette.text.primary,
          fontWeight: "normal",
          backgroundColor: "transparent",
          position: "relative",
          overflow: "hidden",
          pb: 0,
          "&:hover": {
            opacity: 0.8,
          },
        },
        containerStyles: {
          display: "inline-block",
          position: "relative" as const,
        },
        externalIconStyles: {
          position: "relative" as const,
          marginLeft: 1,
          fontSize: 14,
          color: theme.palette.text.secondary,
          opacity: 0.7,
          padding: "1px",
        },
        titleStyles: {
          display: "flex",
          alignItems: "center",
          justifyContent: "center",
          fontSize: "0.875rem",
          whiteSpace: "nowrap" as const,
          overflow: "hidden" as const,
          textOverflow: "ellipsis" as const,
          maxWidth: 100,
        },
      }),
      [theme.palette.text.primary, theme.palette.text.secondary, url.title, url.iconPath],
    );

    return (
      <Box sx={styles.containerStyles}>
        <Tooltip title={tooltipText} placement="bottom" enterDelay={700}>
          <Button
            onClick={handleClick}
            sx={styles.buttonStyles}
            disableRipple={false}
            aria-label={`${url.title} (opens in new tab)`}
          >
            <Box sx={styles.boxStyles}>
              {url.iconPath ? (
                <>
                  <Box component="img" src={url.iconPath} alt="" sx={styles.iconStyles} />
                  {url.title && <Box sx={styles.titleStyles}>{url.title}</Box>}
                </>
              ) : (
                <Box sx={styles.titleStyles}>{url.title}</Box>
              )}
              <OpenInNewIcon sx={styles.externalIconStyles} fontSize="small" />
            </Box>
          </Button>
        </Tooltip>
      </Box>
    );
  }

  // For side menu, use ListItemButton-based styling to match UrlItem
  return (
    <ListItemButton
      onClick={handleClick}
      sx={{
        pl: isMobile ? 3 : 4,
        position: "relative",
        borderRight: 0,
        display: "flex",
        alignItems: "center",
      }}
    >
      {url.iconPath && (
        <Box
          component="img"
          src={url.iconPath}
          alt=""
          sx={{
            width: 20,
            height: 20,
            objectFit: "contain",
            mr: 1.5,
            flexShrink: 0,
          }}
        />
      )}
      <ListItemText
        primary={url.title || url.url}
        secondary={url.url}
        primaryTypographyProps={{
          fontSize: isMobile ? "0.875rem" : "inherit", // Smaller font on mobile
        }}
        secondaryTypographyProps={{
          fontSize: isMobile ? "0.75rem" : "inherit", // Smaller font on mobile
        }}
      />
      <OpenInNewIcon
        sx={{
          ml: 1,
          fontSize: isMobile ? 12 : 14, // Smaller icon on mobile
          color: theme.palette.text.secondary,
          opacity: 0.7,
        }}
        fontSize="small"
      />
    </ListItemButton>
  );
});

export { ExternalUrlItem };
