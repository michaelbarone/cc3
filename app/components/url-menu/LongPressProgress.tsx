import { Box, LinearProgress, useTheme } from "@mui/material";

interface LongPressProgressProps {
  progress: number;
  isActive: boolean;
}

export function LongPressProgress({ progress, isActive }: LongPressProgressProps) {
  const theme = useTheme();

  if (!isActive) return null;

  // Using orange color as specified in requirements
  const progressColor = theme.palette.warning.main;

  return (
    <Box
      sx={{
        position: "absolute",
        left: 0,
        right: 0,
        bottom: 0,
        height: 3, // 3px thick as per requirements (2-3px)
        overflow: "hidden",
        zIndex: 1,
      }}
    >
      <LinearProgress
        variant="determinate"
        value={progress < 25 ? 0 : progress} // 0.3s delay implemented via threshold
        sx={{
          height: "100%",
          backgroundColor: "transparent",
          "& .MuiLinearProgress-bar": {
            backgroundColor: progressColor,
            transition: "none", // Remove default transition for smoother updates
          },
        }}
      />
    </Box>
  );
}
