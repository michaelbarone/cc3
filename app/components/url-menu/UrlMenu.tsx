"use client";

import { useUrlManager } from "@/app/lib/hooks/useIframe";
import { useLongPress } from "@/app/lib/hooks/useLongPress";
import type { UrlGroup } from "@/app/types/iframe";
import { ExpandLess, ExpandMore } from "@mui/icons-material";
import { Box, Collapse, List, ListItem, ListItemButton, ListItemText } from "@mui/material";
import {
  forwardRef,
  memo,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
  type ChangeEvent,
} from "react";
import { LongPressProgress } from "./LongPressProgress";

interface UrlMenuProps {
  urlGroups: UrlGroup[];
  initialUrlId?: string;
  onUrlSelect?: (urlId: string) => void;
}

// Extract UrlMenuItem to a separate memoized component
const UrlMenuItem = memo(
  ({
    url,
    isActive,
    isLoaded,
    onUrlClick,
    onLongPress,
  }: {
    url: UrlGroup["urls"][0] & { title: string };
    isActive: boolean;
    isLoaded: boolean;
    onUrlClick: (urlId: string) => void;
    onLongPress: (urlId: string) => void;
  }) => {
    // Use the hook directly at the top level, not inside useMemo
    const longPressHandlers = useLongPress({
      onClick: () => onUrlClick(url.id),
      onLongPress: () => onLongPress(url.id),
      duration: 2000,
      disabled: false,
      visualFeedback: true,
    });

    // Convert legacy icon paths to API paths if needed
    const iconPath = url.iconPath
      ? url.iconPath.startsWith("/api/public/")
        ? url.iconPath
        : url.iconPath.startsWith("/icons/")
          ? `/api/public${url.iconPath}`
          : url.iconPath
      : null;

    return (
      <ListItem
        disablePadding
        data-url-id={url.id}
        sx={{
          position: "relative",
          "--loaded-indicator": isLoaded ? "''" : "none",
          "&::after": isLoaded
            ? {
                content: "var(--loaded-indicator)",
                position: "absolute",
                top: "50%",
                right: 8,
                transform: "translateY(-50%)",
                width: 8,
                height: 8,
                borderRadius: "50%",
                backgroundColor: "success.main",
                opacity: 0.8,
              }
            : undefined,
        }}
      >
        <ListItemButton
          selected={isActive}
          data-url-id={url.id}
          {...longPressHandlers}
          sx={{
            pl: 4,
            position: "relative",
            borderRight: isActive ? 3 : 0,
            borderColor: "primary.main",
          }}
        >
          {iconPath && (
            <Box
              component="img"
              src={iconPath}
              alt=""
              sx={{
                width: 20,
                height: 20,
                objectFit: "contain",
                mr: 1.5,
                flexShrink: 0,
              }}
            />
          )}
          <ListItemText
            primary={url.title || url.url}
            secondary={url.url}
            primaryTypographyProps={{
              fontWeight: isActive ? "bold" : "normal",
            }}
          />
          <LongPressProgress
            progress={longPressHandlers.progress}
            isActive={longPressHandlers.isLongPressing}
          />
        </ListItemButton>
      </ListItem>
    );
  },
);

// Extract GroupHeader into a memoized component
const GroupHeader = memo(
  forwardRef<
    HTMLDivElement,
    {
      group: UrlGroup;
      isExpanded: boolean;
      onToggle: (groupId: string) => void;
    }
  >((props, ref) => {
    const { group, isExpanded, onToggle } = props;

    const handleClick = useCallback(() => {
      onToggle(group.id);
    }, [onToggle, group.id]);

    return (
      <ListItemButton component="div" ref={ref} onClick={handleClick} data-group-id={group.id}>
        <ListItemText primary={group.name} />
        {isExpanded ? <ExpandLess /> : <ExpandMore />}
      </ListItemButton>
    );
  }),
);

GroupHeader.displayName = "GroupHeader";
UrlMenuItem.displayName = "UrlMenuItem";

export function UrlMenu({ urlGroups, initialUrlId, onUrlSelect }: UrlMenuProps) {
  // State
  const [expandedGroups, setExpandedGroups] = useState<Set<string>>(new Set());
  const [searchQuery, setSearchQuery] = useState("");
  const searchInputRef = useRef<HTMLInputElement>(null);
  const firstGroupHeaderRef = useRef<HTMLDivElement>(null);

  // Use useEffect for client-side state initialization
  const [mounted, setMounted] = useState(false);
  useEffect(() => {
    setMounted(true);
  }, []);

  const { urls, activeUrlId, selectUrl, unloadUrl } = useUrlManager(urlGroups);

  // Initialize URLs if needed
  useEffect(() => {
    if (initialUrlId && mounted) {
      selectUrl(initialUrlId);
    }
  }, [initialUrlId, selectUrl, mounted]);

  // Handle URL selection
  const handleUrlClick = useCallback(
    (urlId: string) => {
      selectUrl(urlId);
      onUrlSelect?.(urlId);
    },
    [selectUrl, onUrlSelect],
  );

  // Handle URL unload (long press)
  const handleUrlUnload = useCallback(
    (urlId: string) => {
      unloadUrl(urlId);
    },
    [unloadUrl],
  );

  // Find the group containing the active URL
  const activeGroupId = useMemo(() => {
    if (!activeUrlId) return null;

    const group = urlGroups.find((group) => group.urls.some((url) => url.id === activeUrlId));

    return group?.id || null;
  }, [activeUrlId, urlGroups]);

  // Ensure the active group is expanded when the active URL changes
  useEffect(() => {
    if (activeGroupId) {
      setExpandedGroups((prev) => {
        const next = new Set(prev);
        next.add(activeGroupId);
        return next;
      });
    }
  }, [activeGroupId]);

  // Modify the handleGroupToggle function to prevent collapsing the active group
  const handleGroupToggle = useCallback(
    (groupId: string) => {
      setExpandedGroups((prev) => {
        const next = new Set(prev);

        // If this is the active group, don't allow it to be collapsed
        // if (groupId === activeGroupId) {
        //   next.add(groupId);
        //   return next;
        // }

        // Otherwise toggle as normal
        if (next.has(groupId)) {
          next.delete(groupId);
        } else {
          next.add(groupId);
        }
        return next;
      });
    },
    [activeGroupId],
  );

  // Handle search input
  const handleSearchChange = useCallback((event: ChangeEvent<HTMLInputElement>) => {
    setSearchQuery(event.target.value.toLowerCase());
  }, []);

  // Filter URLs based on search query
  const filteredGroups = useMemo(() => {
    if (!searchQuery) return urlGroups;

    return urlGroups
      .map((group) => ({
        ...group,
        urls: group.urls.filter(
          (url) =>
            url.url.toLowerCase().includes(searchQuery) ||
            (url.title || "").toLowerCase().includes(searchQuery),
        ),
      }))
      .filter((group) => group.urls.length > 0);
  }, [urlGroups, searchQuery]);

  // Focus search input on keyboard shortcut
  useEffect(() => {
    if (!mounted) return;

    const handleKeyDown = (event: KeyboardEvent) => {
      if ((event.ctrlKey || event.metaKey) && event.key === "k") {
        event.preventDefault();
        searchInputRef.current?.focus();
      }
    };

    window.addEventListener("keydown", handleKeyDown);
    return () => window.removeEventListener("keydown", handleKeyDown);
  }, [mounted]);

  // Prevent hydration mismatch by not rendering anything on the server
  if (!mounted) {
    return (
      <Box sx={{ width: "100%", height: "100%", display: "flex", flexDirection: "column" }}>
        <List
          sx={{
            width: "100%",
            flex: 1,
            overflowY: "auto",
            bgcolor: "background.paper",
          }}
          component="nav"
        />
      </Box>
    );
  }

  return (
    <Box sx={{ width: "100%", height: "100%", display: "flex", flexDirection: "column" }}>
      {/* Search functionality is commented out for now */}
      {/* <Box sx={{ p: 1 }}>
        <TextField
          fullWidth
          size="small"
          inputRef={searchInputRef}
          placeholder="Search URLs (Ctrl+K)"
          value={searchQuery}
          onChange={handleSearchChange}
          InputProps={{
            startAdornment: <SearchIcon sx={{ mr: 1, color: "text.secondary" }} />,
          }}
        />
      </Box> */}

      <List
        sx={{
          width: "100%",
          flex: 1,
          overflowY: "auto",
          bgcolor: "background.paper",
        }}
        component="nav"
      >
        {filteredGroups.length === 0 ? (
          <></>
        ) : (
          filteredGroups.map((group) => (
            <Box key={group.id}>
              <GroupHeader
                ref={group === filteredGroups[0] ? firstGroupHeaderRef : undefined}
                group={group}
                isExpanded={expandedGroups.has(group.id)}
                onToggle={handleGroupToggle}
              />
              <Collapse in={expandedGroups.has(group.id)} timeout="auto" unmountOnExit>
                <List component="div" disablePadding aria-label={`${group.name} URLs`}>
                  {group.urls.map((url) => (
                    <UrlMenuItem
                      key={url.id}
                      url={{ ...url, title: url.title || url.url }}
                      isActive={url.id === activeUrlId}
                      isLoaded={urls[url.id]?.isLoaded ?? false}
                      onUrlClick={handleUrlClick}
                      onLongPress={handleUrlUnload}
                    />
                  ))}
                </List>
              </Collapse>
            </Box>
          ))
        )}
      </List>
    </Box>
  );
}
