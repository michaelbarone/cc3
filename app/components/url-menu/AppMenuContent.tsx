"use client";

import { useUrlGroups } from "@/app/lib/hooks/useUrlGroups";
import { useUserPreferences } from "@/app/lib/hooks/useUserPreferences";
import { useMediaQuery, useTheme } from "@mui/material";
import { TopMenuNavigation } from "./TopMenuNavigation";
import { UrlMenu } from "./UrlMenu";

interface AppMenuContentProps {
  forceMenuPosition?: "side" | "top" | null;
  initialUrlId?: string | null;
}

export function AppMenuContent({
  forceMenuPosition = null,
  initialUrlId = null,
}: AppMenuContentProps) {
  const theme = useTheme();
  const { preferences } = useUserPreferences();
  const isMobile = useMediaQuery(theme.breakpoints.down("md"));
  const { urlGroups, isLoading, error } = useUrlGroups();

  // Determine effective menu position based on user preferences and device
  const effectiveMenuPosition = (() => {
    // Mobile always uses side menu
    if (isMobile) return "side";

    // Use forced position if provided
    if (forceMenuPosition) return forceMenuPosition;

    // Use user preference or default to "top"
    return preferences?.menuPosition || "top";
  })();

  // Handle loading and error states
  if (isLoading) return null;
  if (error) return null;
  if (!urlGroups || urlGroups.length === 0) return null;

  // Render menu based on effective position
  if (effectiveMenuPosition === "top" && !isMobile) {
    return <TopMenuNavigation urlGroups={urlGroups} initialUrlId={initialUrlId || undefined} />;
  }

  // Default to side menu
  return <UrlMenu urlGroups={urlGroups} initialUrlId={initialUrlId || undefined} />;
}
