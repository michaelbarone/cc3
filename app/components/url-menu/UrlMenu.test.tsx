import { UrlMenu } from "@/app/components/url-menu/UrlMenu";
import { IframeProvider } from "@/app/lib/state/iframe-state";
import { measureTestTime, THRESHOLDS } from "@/test/helpers/debug";
import { fireEvent, render, screen, waitFor, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { ReactNode } from "react";
import { act } from "react-dom/test-utils";
import { afterAll, beforeEach, describe, expect, it, vi } from "vitest";

/**
 * UrlMenu Component Tests
 *
 * This test suite includes standard functionality tests and edge case tests:
 *
 * Edge Case Tests:
 * 1. Rapid Clicks: Tests rapid sequential clicks between different URL buttons
 *    to ensure proper state management and that the last click is properly registered.
 *
 * 2. Interrupted Long Press: Tests that long press actions are properly canceled
 *    when interrupted before the threshold duration (via mouse movement or early release).
 *
 * 3. Browser Tab Switching: Tests that operations (especially long press) handle
 *    focus/blur events correctly when a user switches browser tabs during the operation.
 *    This ensures proper state cleanup and prevention of unintended side effects.
 *
 * These edge case tests help ensure the component is robust in real-world usage scenarios
 * where user interactions may be unpredictable or interrupted.
 */

// Mock data
const mockUrlGroups = [
  {
    id: "group1",
    name: "Group 1",
    urls: [
      {
        id: "url1",
        title: "URL 1",
        url: "https://example.com/1",
        urlMobile: null,
        iconPath: null,
        idleTimeoutMinutes: null,
      },
      {
        id: "url2",
        title: "URL 2",
        url: "https://example.com/2",
        urlMobile: null,
        iconPath: null,
        idleTimeoutMinutes: null,
      },
    ],
  },
  {
    id: "group2",
    name: "Group 2",
    urls: [
      {
        id: "url3",
        title: "URL 3",
        url: "https://example.com/3",
        urlMobile: null,
        iconPath: null,
        idleTimeoutMinutes: null,
      },
      {
        id: "url4",
        title: "URL 4",
        url: "https://example.com/4",
        urlMobile: null,
        iconPath: null,
        idleTimeoutMinutes: null,
      },
    ],
  },
];

// Updated TestWrapper to use IframeProvider
const TestWrapper = ({ children }: { children: ReactNode }) => {
  return <IframeProvider>{children}</IframeProvider>;
};

describe("UrlMenu", () => {
  const suiteTimer = measureTestTime("UrlMenu Suite");

  beforeEach(() => {
    vi.clearAllMocks();
  });

  afterAll(() => {
    suiteTimer.end();
  });

  const renderUrlMenu = (props = {}) => {
    return render(
      <TestWrapper>
        <UrlMenu urlGroups={mockUrlGroups} {...props} />
      </TestWrapper>,
    );
  };

  const expandGroup = async (groupName: string) => {
    const groupHeader = screen.getByText(groupName).closest("[data-group-id]") as HTMLElement;
    fireEvent.click(groupHeader);
    await waitFor(
      () => {
        const list = screen.getByRole("list", { name: `${groupName} URLs` });
        expect(list).toBeInTheDocument();
      },
      { timeout: 2000 },
    );
    return groupHeader;
  };

  it("should render all groups", () => {
    const testTimer = measureTestTime("render all groups");
    try {
      renderUrlMenu();
      expect(screen.getByText("Group 1")).toBeInTheDocument();
      expect(screen.getByText("Group 2")).toBeInTheDocument();
      expect(testTimer.elapsed()).toBeLessThan(THRESHOLDS.UNIT);
    } finally {
      testTimer.end();
    }
  });

  it("should expand/collapse groups", async () => {
    const testTimer = measureTestTime("expand/collapse groups");
    try {
      renderUrlMenu();

      // Initially all groups are collapsed
      const group1Header = screen.getByText("Group 1").closest("[data-group-id]") as HTMLElement;
      expect(group1Header).toBeInTheDocument();

      // Click to expand
      fireEvent.click(group1Header);

      // Wait for list to be visible
      await waitFor(() => {
        const list = screen.getByRole("list", { name: "Group 1 URLs" });
        expect(list).toBeInTheDocument();
        expect(within(list).getByText("URL 1")).toBeInTheDocument();
        expect(within(list).getByText("URL 2")).toBeInTheDocument();
      });

      // Click to collapse
      fireEvent.click(group1Header);

      // Wait for list to be hidden
      await waitFor(() => {
        expect(screen.queryByRole("list", { name: "Group 1 URLs" })).not.toBeInTheDocument();
      });
      expect(testTimer.elapsed()).toBeLessThan(THRESHOLDS.UNIT);
    } finally {
      testTimer.end();
    }
  });

  it("should handle keyboard navigation", async () => {
    const testTimer = measureTestTime("handle keyboard navigation");
    try {
      renderUrlMenu();
      const user = userEvent.setup();

      // Focus first group header
      const group1Header = screen.getByRole("button", { name: "Group 1" });
      await user.click(group1Header);

      // Wait for the animation to complete and the list to be visible
      await waitFor(
        () => {
          const list = screen.queryByRole("list", { name: "Group 1 URLs" });
          expect(list).toBeInTheDocument();
        },
        { timeout: 2000 },
      );

      // Press / to focus search
      await user.keyboard("/");
      expect(document.activeElement?.getAttribute("placeholder")).toBe("Search URLs");

      // Press Escape to return focus to group header
      await user.keyboard("{Escape}");
      expect(document.activeElement).toBe(group1Header);
      expect(testTimer.elapsed()).toBeLessThan(THRESHOLDS.UNIT);
    } finally {
      testTimer.end();
    }
  });

  it("should handle URL selection", async () => {
    const testTimer = measureTestTime("handle URL selection");
    try {
      const onUrlSelect = vi.fn();
      renderUrlMenu({ onUrlSelect });

      // Expand group 1
      const group1Header = screen.getByRole("button", { name: "Group 1" });
      await userEvent.click(group1Header);

      // Wait for the animation to complete
      await waitFor(
        () => {
          const list = screen.queryByRole("list", { name: "Group 1 URLs" });
          expect(list).toBeInTheDocument();
        },
        { timeout: 1000 },
      );

      // Find the URL button
      const urlButton = screen.getByRole("button", { name: /URL 1/ });

      // Click the URL button
      await userEvent.click(urlButton);

      // Verify the click handler was called
      await waitFor(
        () => {
          expect(onUrlSelect).toHaveBeenCalledWith("url1");
        },
        { timeout: 1000 },
      );
      expect(testTimer.elapsed()).toBeLessThan(THRESHOLDS.UNIT);
    } finally {
      testTimer.end();
    }
  });

  it("should handle URL state transitions", async () => {
    const testTimer = measureTestTime("handle URL state transitions");
    try {
      renderUrlMenu({ activeUrlId: "url1" });

      // Expand group 1
      await expandGroup("Group 1");

      // Check URL 1 is selected
      const list = screen.getByRole("list", { name: "Group 1 URLs" });
      const url1Button = within(list).getByRole("button", { name: /URL 1/ });
      expect(url1Button).toHaveClass("Mui-selected");
      expect(testTimer.elapsed()).toBeLessThan(THRESHOLDS.UNIT);
    } finally {
      testTimer.end();
    }
  });

  it("should persist group collapse state", async () => {
    const testTimer = measureTestTime("persist group collapse state");
    try {
      renderUrlMenu();

      // Expand group 1
      await expandGroup("Group 1");

      // Check URLs are visible
      const list = screen.getByRole("list", { name: "Group 1 URLs" });
      const urlItems = within(list).getAllByRole("listitem");
      expect(urlItems).toHaveLength(2);
      expect(testTimer.elapsed()).toBeLessThan(THRESHOLDS.UNIT);
    } finally {
      testTimer.end();
    }
  });

  it("should handle search filtering", async () => {
    const testTimer = measureTestTime("handle search filtering");
    try {
      renderUrlMenu();

      // Expand both groups
      await expandGroup("Group 1");
      await expandGroup("Group 2");

      // Type in search
      const searchInput = screen.getByPlaceholderText("Search URLs");
      await userEvent.type(searchInput, "URL 1");

      // Check filtered results
      await waitFor(() => {
        const group1List = screen.getByRole("list", { name: "Group 1 URLs" });
        expect(within(group1List).getByText("URL 1")).toBeInTheDocument();
        expect(within(group1List).queryByText("URL 2")).not.toBeInTheDocument();
        expect(screen.queryByRole("list", { name: "Group 2 URLs" })).not.toBeInTheDocument();
      });
      expect(testTimer.elapsed()).toBeLessThan(THRESHOLDS.UNIT);
    } finally {
      testTimer.end();
    }
  });

  it("should handle empty states", () => {
    const testTimer = measureTestTime("handle empty states");
    try {
      render(
        <TestWrapper>
          <UrlMenu urlGroups={[]} />
        </TestWrapper>,
      );

      expect(screen.getByText("No URLs available")).toBeInTheDocument(); // Update based on your actual empty state message
      expect(testTimer.elapsed()).toBeLessThan(THRESHOLDS.UNIT);
    } finally {
      testTimer.end();
    }
  });

  it("should handle long press to unload URL", async () => {
    const testTimer = measureTestTime("handle long press to unload URL");
    try {
      const user = userEvent.setup();
      render(
        <TestWrapper>
          <UrlMenu urlGroups={mockUrlGroups} />
        </TestWrapper>,
      );

      // Expand Group 1
      const group1Header = screen.getByRole("button", { name: "Group 1" });
      await user.click(group1Header);

      // Wait for list to be visible and animation to complete
      await waitFor(() => {
        const list = screen.getByRole("list", { name: "Group 1 URLs" });
        expect(list).toBeInTheDocument();
        expect(list.closest(".MuiCollapse-root")).toHaveClass("MuiCollapse-entered");
      });

      // Find the URL button
      const urlItem = screen.getByText("URL 1").closest("li");
      const urlButton = within(urlItem!).getByRole("button");

      // Simulate long press by firing mouseDown and waiting
      await act(async () => {
        fireEvent.mouseDown(urlButton);
        await new Promise((resolve) => setTimeout(resolve, 600)); // Wait longer than the 500ms threshold
        fireEvent.mouseUp(urlButton);
      });

      // Check for unload effect (can be checked through classes or aria-attributes based on your implementation)
      // This needs to be updated based on how your component shows unload state
      expect(urlButton).toHaveAttribute("aria-pressed", "false");
      expect(testTimer.elapsed()).toBeLessThan(THRESHOLDS.UNIT);
    } finally {
      testTimer.end();
    }
  });

  // New test for rapid clicks between different URLs
  it("should handle rapid clicks between different URLs", async () => {
    const testTimer = measureTestTime("handle rapid clicks between different URLs");
    try {
      const onUrlSelect = vi.fn();
      renderUrlMenu({ onUrlSelect });

      // Expand group 1
      await expandGroup("Group 1");

      // Find URL buttons
      const url1Button = screen.getByRole("button", { name: /URL 1/ });
      const url2Button = screen.getByRole("button", { name: /URL 2/ });

      // Rapid clicks between URLs
      await act(async () => {
        fireEvent.click(url1Button);
        fireEvent.click(url2Button);
        fireEvent.click(url1Button);
        fireEvent.click(url2Button);
        fireEvent.click(url1Button);
      });

      // Verify the last click was registered correctly
      await waitFor(() => {
        expect(onUrlSelect).toHaveBeenLastCalledWith("url1");
        expect(onUrlSelect).toHaveBeenCalledTimes(5);
      });
      expect(testTimer.elapsed()).toBeLessThan(THRESHOLDS.UNIT);
    } finally {
      testTimer.end();
    }
  });

  // New test for interrupted long press behavior
  it("should handle interrupted long press correctly", async () => {
    const testTimer = measureTestTime("handle interrupted long press correctly");
    try {
      const user = userEvent.setup();
      renderUrlMenu();

      // Expand Group 1
      await expandGroup("Group 1");

      // Find the URL button
      const urlItem = screen.getByText("URL 1").closest("li");
      const urlButton = within(urlItem!).getByRole("button");

      // Start long press but interrupt before threshold
      await act(async () => {
        fireEvent.mouseDown(urlButton);
        // Wait less than the long press threshold (500ms)
        await new Promise((resolve) => setTimeout(resolve, 250));
        // Interrupt by moving mouse
        fireEvent.mouseMove(urlButton, { clientX: 10, clientY: 10 });
        fireEvent.mouseUp(urlButton);
      });

      // Verify no long press action occurred
      // This depends on your implementation - check a specific property or state
      // that indicates long press was not completed
      expect(urlButton).not.toHaveAttribute("data-long-press-active", "true");
      expect(urlButton).toHaveAttribute("aria-pressed", "false");
      expect(testTimer.elapsed()).toBeLessThan(THRESHOLDS.UNIT);
    } finally {
      testTimer.end();
    }
  });

  // New test for browser tab switching during long operations
  it("should handle browser tab switching during operations", async () => {
    const testTimer = measureTestTime("handle browser tab switching during operations");
    try {
      renderUrlMenu();

      // Expand Group 1
      await expandGroup("Group 1");

      // Find the URL button
      const urlItem = screen.getByText("URL 1").closest("li");
      const urlButton = within(urlItem!).getByRole("button");

      // Start long press
      await act(async () => {
        fireEvent.mouseDown(urlButton);

        // Simulate tab losing focus during long press
        fireEvent.blur(window);

        // Wait enough time for long press to trigger
        await new Promise((resolve) => setTimeout(resolve, 600));

        // Tab gaining focus again
        fireEvent.focus(window);

        fireEvent.mouseUp(urlButton);
      });

      // Verify the long press was canceled when focus was lost
      expect(urlButton).not.toHaveAttribute("data-long-press-active", "true");
      expect(urlButton).toHaveAttribute("aria-pressed", "false");
      expect(testTimer.elapsed()).toBeLessThan(THRESHOLDS.UNIT);
    } finally {
      testTimer.end();
    }
  });
});
