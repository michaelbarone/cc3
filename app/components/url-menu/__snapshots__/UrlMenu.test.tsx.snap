// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UrlMenu > should handle URL state transitions 1`] = `
<li
  class="MuiListItem-root MuiListItem-gutters css-meaawr-MuiListItem-root"
  data-url-id="url1"
>
  <div
    class="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters Mui-selected MuiListItemButton-root MuiListItemButton-gutters Mui-selected css-dgraio-MuiButtonBase-root-MuiListItemButton-root"
    role="button"
    tabindex="0"
  >
    <div
      class="MuiListItemText-root MuiListItemText-multiline css-14ln1j6-MuiListItemText-root"
    >
      <span
        class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-lhsbas-MuiTypography-root"
      >
        URL 1
      </span>
      <p
        class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-1a1whku-MuiTypography-root"
      >
        https://example.com/1
      </p>
    </div>
  </div>
</li>
`;
